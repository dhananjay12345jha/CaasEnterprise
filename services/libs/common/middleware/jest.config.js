module.exports = {
  displayName: "common-middleware",
  preset: "../../../tools/jest.preset.js",
  silent: true, // Silencing logs on the stdout. Set it to false if you need console.logs
  globals: {
    "ts-jest": {
      tsconfig: "<rootDir>/tsconfig.spec.json",
    },
  },
  testEnvironment: "node",
  setupFiles: ["<rootDir>/dotenv-config.ts"],
  transform: {
    "^.+\\.[tj]sx?$": "ts-jest",
  },
  moduleFileExtensions: ["ts", "tsx", "js", "jsx", "json"],
  coverageDirectory: "../../../../coverage/libs/common/middleware",
  reporters: [
    "default",
    [
      "jest-junit",
      { outputDirectory: "./junit", outputName: "serverless-middleware.xml" },
    ],
  ],
  coverageThreshold: {
    global: {
      branches: 84,
      functions: 85,
      lines: 85,
      statements: 85,
    },
  },
};
