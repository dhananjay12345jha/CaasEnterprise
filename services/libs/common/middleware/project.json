{
  "root": "libs/common/middleware",
  "sourceRoot": "libs/common/middleware/src",
  "projectType": "library",
  "targets": {
    "lint": {
      "executor": "@nrwl/linter:eslint",
      "outputs": ["{options.outputFile}"],
      "options": {
        "lintFilePatterns": ["libs/common/middleware/**/*.ts"]
      }
    },
    "test": {
      "executor": "@nrwl/jest:jest",
      "outputs": ["coverage/libs/common/middleware"],
      "options": {
        "jestConfig": "libs/common/middleware/jest.config.js",
        "passWithNoTests": true
      }
    }
  },
  "tags": ["type:lib"]
}
