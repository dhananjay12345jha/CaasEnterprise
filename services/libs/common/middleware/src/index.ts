export * from "./lib/new-relic";
export * from "./lib/validator";
export * from "./lib/request-schema";
export * from "./lib/sqs";
