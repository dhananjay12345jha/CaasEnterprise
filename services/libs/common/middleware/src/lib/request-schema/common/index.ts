import { z } from "../../validator";
import { ZodTypeAny } from "zod";

export const buildEventMetaSchema = <T extends string>(detailType: T) =>
  z.object({
    version: z.string().min(1),
    id: z.string().min(1),
    "detail-type": z.literal(detailType),
    source: z.string().optional(),
    account: z.string().optional(),
    time: z.string().min(1),
    region: z.string().optional(),
    resources: z.array(z.string()).optional(),
  });

export const metadataSchema = z.object({
  "x-emc-ubid": z.string().min(1),
  "x-amzn-RequestId": z.string().optional(),
  "x-lambda-RequestId": z.string().optional(),
});

export const detailSchema = z.object({
  referenced: z.boolean().optional(),
  correlationId: z.string().uuid().optional(),
});

export const eventDetailSchema = <T extends ZodTypeAny>(payloadSchema: T) =>
  detailSchema
    .extend({ metadata: metadataSchema })
    .extend({ payload: payloadSchema });

export const buildEventSchema = <S extends string, T extends ZodTypeAny>(
  detailType: S,
  payloadSchema: T
) =>
  buildEventMetaSchema(detailType).extend({
    detail: eventDetailSchema(payloadSchema),
  });
