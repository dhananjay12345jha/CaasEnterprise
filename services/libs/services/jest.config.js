module.exports = {
  displayName: "common-services",
  preset: "../../tools/jest.preset.js",
  globals: {
    "ts-jest": {
      tsconfig: "<rootDir>/tsconfig.spec.json",
    },
  },
  testEnvironment: "node",
  transform: {
    "^.+\\.[tj]sx?$": "ts-jest",
  },
  moduleFileExtensions: ["ts", "tsx", "js", "jsx"],
  coverageDirectory: "../../coverage/libs/services",
  coverageReporters: ["text", "cobertura", "json-summary", "html"],
  reporters: [
    "default",
    ["jest-junit", { outputDirectory: "./junit", outputName: "services.xml" }],
  ],
  coverageThreshold: {
    global: {
      branches: 85,
      functions: 85,
      lines: 85,
      statements: 85,
    },
  },
};
