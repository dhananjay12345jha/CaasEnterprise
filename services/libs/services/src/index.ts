export * from "./lib/errors/";
export * from "./lib/interfaces/";
export * from "./lib/psk/";
export * from "./lib/carts/";
export * from "./lib/orders/";
export * from "./lib/inventory/";
export * from "./lib/crm";
export * from "./lib/logger/";
export * from "./lib/payments/";
