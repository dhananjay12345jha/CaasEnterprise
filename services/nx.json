{
  "extends": "@nrwl/workspace/presets/core.json",
  "npmScope": "ei-services",
  "affected": {
    "defaultBase": "origin/develop"
  },
  "workspaceLayout": {
    "appsDir": "apps",
    "libsDir": "libs"
  },
  "cli": {
    "defaultCollection": "@nrwl/js",
    "packageManager": "yarn"
  },
  "tasksRunnerOptions": {
    "default": {
      "runner": "@nrwl/workspace/tasks-runners/default",
      "options": {
        "cacheableOperations": ["build", "lint", "test", "e2e"],
        "runtimeCacheInputs": [
          "echo BUILD_CONTEXT=${BUILD_CONTEXT:=local}",
          "echo DELIVERY_CONTEXT=${DELIVERY_CONTEXT:=sandbox}",
          "echo DELIVERY_STAGE=${DELIVERY_STAGE:=dev}"
        ]
      }
    }
  },
  "targetDefaults": {
    "build": {
      "dependsOn": ["^build"]
    }
  },
  "implicitDependencies": {
    "workspace.json": "*",
    "package.json": {
      "dependencies": "*",
      "devDependencies": "*"
    },
    "tsconfig.base.json": "*",
    ".eslintrc.json": "*",
    "nx.json": "*"
  }
}
