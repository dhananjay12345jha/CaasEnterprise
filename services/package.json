{
  "name": "@enterprise-integration/services",
  "version": "0.0.0",
  "private": true,
  "scripts": {
    "nx": "nx",
    "lint": "yarn nx workspace-lint && yarn nx lint",
    "build": "yarn nx build",
    "test": "yarn nx test",
    "test:unit": "yarn run test",
    "e2e": "yarn nx e2e",
    "format": "yarn nx format:write",
    "format:write": "yarn nx format:write",
    "format:check": "yarn nx format:check",
    "dep-graph": "yarn nx dep-graph",
    "serve": "yarn nx serve",
    "serve:dev": "yarn nx run-many --all --target=serve:local",
    "affected": "yarn nx affected",
    "affected:apps": "yarn nx affected:apps",
    "affected:libs": "yarn nx affected:libs",
    "affected:lint": "yarn nx affected:lint",
    "affected:info": "yarn nx affected --target=info",
    "affected:build": "yarn nx affected:build",
    "affected:test:feature": "yarn affected --target=e2e-feature",
    "affected:test:unit": "yarn nx affected:test",
    "affected:test:component": "yarn nx affected:e2e",
    "affected:test:performance": "yarn affected --target=insprint-performance",
    "affected:test:integration": "yarn affected --target=e2e-integration",
    "affected:version": "yarn version --patch --deferred",
    "affected:version:check": "yarn version check",
    "affected:version:apply": "yarn version apply",
    "affected:deploy": "yarn nx affected --target=deploy",
    "affected:destroy": "yarn nx affected --target=destroy",
    "workspaces:lint": "yarn nx workspace-lint"
  },
  "dependencies": {
    "@adyen/api-library": "^12.0.0",
    "@aws-sdk/client-dynamodb": "^3.105.0",
    "@aws-sdk/client-eventbridge": "^3.105.0",
    "@aws-sdk/client-sqs": "^3.112.0",
    "@aws-sdk/util-dynamodb": "^3.105.0",
    "@commercetools/platform-sdk": "^2.7.0",
    "@commercetools/sdk-client-v2": "^1.4.0",
    "@everymile-idp/brand-config-sdk": "file:../.idp-shim/dist/brand-config-sdk",
    "@everymile-schemas/order-created": "file:../contracts/private/jsonSchema/order/created",
    "@middy/core": "^2.5.3",
    "@middy/http-json-body-parser": "^2.5.3",
    "@middy/validator": "^3.1.0",
    "@types/fnv-plus": "^1.3.0",
    "ajv": "^8.11.0",
    "ajv-draft-04": "^1.0.0",
    "ajv-formats": "^2.1.1",
    "aws-lambda": "^1.0.7",
    "axios": "^0.27.2",
    "axios-retry": "^3.3.1",
    "fnv-plus": "^1.3.1",
    "iso-country-currency": "^0.6.0",
    "iso-datestring-validator": "^2.2.2",
    "newrelic": "^8.13.1",
    "node-fetch": "^2.6.1",
    "traverse": "^0.6.6",
    "tslib": "^2.4.0",
    "winston": "^3.8.2",
    "zod": "^3.17.3"
  },
  "devDependencies": {
    "@aws-sdk/client-eventbridge": "^3.110.0",
    "@enterprise-integration/mocks": "workspace:*",
    "@faker-js/faker": "^7.2.0",
    "@nrwl/cli": "13.10.1",
    "@nrwl/cypress": "13.10.1",
    "@nrwl/eslint-plugin-nx": "13.10.1",
    "@nrwl/jest": "13.10.1",
    "@nrwl/js": "13.10.1",
    "@nrwl/linter": "13.10.1",
    "@nrwl/node": "13.10.1",
    "@nrwl/tao": "13.10.1",
    "@nrwl/web": "13.10.1",
    "@nrwl/workspace": "13.10.1",
    "@serverless/typescript": "^3.21.0",
    "@types/aws-lambda": "^8.10.100",
    "@types/jest": "^27.5.1",
    "@types/lodash": "^4",
    "@types/newrelic": "^7.0.3",
    "@types/node": "^18.0.1",
    "@types/node-fetch": "^2.6.1",
    "@types/ramda": "^0.28.15",
    "@types/randomstring": "^1",
    "@types/serverless": "^3.12.7",
    "@types/traverse": "^0.6.32",
    "@types/uuid": "^8",
    "@typescript-eslint/eslint-plugin": "~5.20.0",
    "@typescript-eslint/parser": "~5.18.0",
    "@yarnpkg/esbuild-plugin-pnp": "^3.0.0-rc.2",
    "artillery": "^2.0.0-14",
    "axios": "^0.27.2",
    "axios-mock-adapter": "^1.21.2",
    "cypress": "^9.6.0",
    "dotenv": "^16.0.1",
    "esbuild": "^0.14.11",
    "eslint": "^8.0.0",
    "eslint-config-prettier": "^8.5.0",
    "jest": "27.2.3",
    "jest-fetch-mock": "^3.0.3",
    "jest-junit": "^13.2.0",
    "jest-util": "^28.1.3",
    "json-schema-to-ts": "^1.5.0",
    "kill-port": "^2.0.0",
    "mqtt": "^4.3.7",
    "newrelic": "^9.0.2",
    "nock": "^13.2.6",
    "nx": "13.10.1",
    "prettier": "^2.6.2",
    "ramda": "^0.28.0",
    "randomstring": "^1.2.2",
    "serverless": "^3.21.0",
    "serverless-esbuild": "^1.23.3",
    "serverless-iam-roles-per-function": "^3.2.0",
    "serverless-newrelic-lambda-layers": "^3.3.5",
    "serverless-offline": "^8.8.0",
    "serverless-offline-aws-eventbridge": "^1.6.6",
    "serverless-prune-plugin": "^2.0.1",
    "serverless-webpack": "^5.7.1",
    "start-server-and-test": "^1.14.0",
    "ts-jest": "27.0.5",
    "ts-loader": "^9.3.0",
    "ts-node": "^10.8.0",
    "tsconfig-paths": "^3.9.0",
    "tsconfig-paths-webpack-plugin": "^3.5.2",
    "typescript": "~4.5.5",
    "uuid": "^8.3.2",
    "webpack": "^5.72.0"
  }
}
