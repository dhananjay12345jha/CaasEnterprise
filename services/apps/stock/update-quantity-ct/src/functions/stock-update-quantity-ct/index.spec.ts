import { v4 as uuid } from "uuid";
import { Context as LambdaContext } from "aws-lambda/handler";

import { default as updateQuantityCT, HandlerEvent } from "./index";
import * as enterpriseIntegrationServices from "@ei-services/services";
import { UPDATE_TYPE } from "@ei-services/common/middleware";

jest.mock("@ei-services/services");

describe("updateQuantityCT handler", () => {
  const mockedGetInventoryBySku = jest.spyOn(
    enterpriseIntegrationServices,
    "getInventoryBySku"
  );
  const mockedUpdateInventoryQuantity = jest.spyOn(
    enterpriseIntegrationServices,
    "updateInventoryQuantity"
  );
  let event: HandlerEvent;
  beforeEach(
    () =>
      (event = {
        version: "0",
        id: uuid(),
        "detail-type": "stock.quantity.updated",
        source: "emc.stock-quantity-updates-sandbox-stockQuantityUpdateWebhook",
        account: "100207810817",
        time: "2022-09-13T17:38:29Z",
        region: "eu-west-1",
        resources: [],
        detail: {
          payload: {
            sku: "handwash-regular",
            quantity: 2,
            updateType: UPDATE_TYPE.INCREMENT,
            time: "2022-06-14T09:35:29.373Z",
          },
          metadata: {
            "x-emc-ubid": "7a733796-ae2c-4032-89f7-24d4c4179612",
            "x-amzn-RequestId": "YaLzHhv8DoEEJJw=",
            "x-lambda-RequestId": "9fd015fc-716f-4481-9dbb-a384489fad52",
          },
        },
      })
  );

  afterEach(() => jest.clearAllMocks());

  describe("when update type is increment", () => {
    it("should not make requests if brand id is missing", async () => {
      event.detail.metadata["x-emc-ubid"] = null;
      try {
        await updateQuantityCT(event, {} as LambdaContext);
        expect("1").toEqual(0);
      } catch (error) {
        expect(error).toMatchSnapshot();
      }
      expect(mockedGetInventoryBySku).not.toHaveBeenCalled();
      expect(mockedUpdateInventoryQuantity).not.toHaveBeenCalled();
    });

    it("should not make requests if sku is missing", async () => {
      event.detail.payload.sku = null;

      await expect(() =>
        updateQuantityCT(event, {} as LambdaContext)
      ).rejects.toMatchSnapshot();
      expect(mockedGetInventoryBySku).not.toHaveBeenCalled();
      expect(mockedUpdateInventoryQuantity).not.toHaveBeenCalled();
    });

    it("should not make requests if quantity is missing", async () => {
      event.detail.payload.quantity = null;
      try {
        await updateQuantityCT(event, {} as LambdaContext);
        expect("1").toEqual(0);
      } catch (error) {
        expect(error).toMatchSnapshot();
      }
      expect(mockedGetInventoryBySku).not.toHaveBeenCalled();
      expect(mockedUpdateInventoryQuantity).not.toHaveBeenCalled();
    });

    it("should not make requests if updateType is missing", async () => {
      event.detail.payload.updateType = null;
      try {
        await updateQuantityCT(event, {} as LambdaContext);
        expect("1").toEqual(0);
      } catch (error) {
        expect(error).toMatchSnapshot();
      }
      expect(mockedGetInventoryBySku).not.toHaveBeenCalled();
      expect(mockedUpdateInventoryQuantity).not.toHaveBeenCalled();
    });

    it("should not make an update request if inventory record is not found", async () => {
      mockedGetInventoryBySku.mockResolvedValueOnce(undefined);
      await updateQuantityCT(event, {} as LambdaContext);
      expect(mockedGetInventoryBySku).toHaveBeenCalled();
      expect(mockedUpdateInventoryQuantity).not.toHaveBeenCalled();
    });

    it("should throw an error if updateInventoryQuantity fails", async () => {
      mockedGetInventoryBySku.mockResolvedValueOnce({
        id: "1",
        version: 1,
        availableQuantity: 1,
        sku: "handwash-regular",
        createdAt: "2021-06-14T09:35:29.373Z",
        lastModifiedAt: "2021-06-14T09:35:29.373Z",
        quantityOnStock: 1,
      });
      mockedUpdateInventoryQuantity.mockRejectedValueOnce(
        new Error("updateInventoryQuantity error")
      );
      await expect(() =>
        updateQuantityCT(event, {} as LambdaContext)
      ).rejects.toThrow("updateInventoryQuantity error");
    });

    it("should call updateInventoryQuantity when event format is correct", async () => {
      mockedGetInventoryBySku.mockResolvedValueOnce({
        id: "1",
        version: 1,
        availableQuantity: 1,
        sku: "handwash-regular",
        createdAt: "2021-06-14T09:35:29.373Z",
        lastModifiedAt: "2021-06-14T09:35:29.373Z",
        quantityOnStock: 1,
      });
      await updateQuantityCT(event, {} as LambdaContext);
      expect(mockedGetInventoryBySku).toHaveBeenCalledTimes(1);
      expect(mockedUpdateInventoryQuantity).toHaveBeenCalledTimes(1);
    });
  });

  describe("when update type is decrement", () => {
    beforeEach(() => {
      event.detail.payload.updateType = UPDATE_TYPE.DECREMENT;
    });

    it("should not call updateInventoryQuantity if event update quantity is greater than inventory entry quantity", async () => {
      mockedGetInventoryBySku.mockResolvedValueOnce({
        id: "1",
        version: 1,
        availableQuantity: 1,
        sku: "handwash-regular",
        createdAt: "2021-06-14T09:35:29.373Z",
        lastModifiedAt: "2021-06-16T09:35:29.373Z",
        quantityOnStock: 1,
      });
      await updateQuantityCT(event, {} as LambdaContext);
      expect(mockedGetInventoryBySku).toHaveBeenCalled();
      expect(mockedUpdateInventoryQuantity).not.toHaveBeenCalled();
    });

    it("should not call updateInventoryQuantity if inventory entry quantity is already 0", async () => {
      event.detail.payload.quantity = 0;
      mockedGetInventoryBySku.mockResolvedValueOnce({
        id: "1",
        version: 1,
        availableQuantity: 0,
        sku: "handwash-regular",
        createdAt: "2021-06-14T09:35:29.373Z",
        lastModifiedAt: "2021-06-16T09:35:29.373Z",
        quantityOnStock: 0,
      });
      await updateQuantityCT(event, {} as LambdaContext);
      expect(mockedGetInventoryBySku).toHaveBeenCalled();
      expect(mockedUpdateInventoryQuantity).not.toHaveBeenCalled();
    });

    it("should call updateInventoryQuantity if event update timestamp is newer than inventory entry last modified timestamp and event update quantity is greater than inventory entry quantity", async () => {
      mockedGetInventoryBySku.mockResolvedValueOnce({
        id: "1",
        version: 1,
        availableQuantity: 3,
        sku: "handwash-regular",
        createdAt: "2021-06-14T09:35:29.373Z",
        lastModifiedAt: "2021-06-12T09:35:29.373Z",
        quantityOnStock: 3,
      });
      await updateQuantityCT(event, {} as LambdaContext);
      expect(mockedGetInventoryBySku).toHaveBeenCalled();
      expect(mockedUpdateInventoryQuantity).toHaveBeenCalled();
    });
  });
});
