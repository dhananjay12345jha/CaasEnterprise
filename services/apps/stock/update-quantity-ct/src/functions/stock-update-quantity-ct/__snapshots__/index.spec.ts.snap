// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`updateQuantityCT handler when update type is increment should not make requests if brand id is missing 1`] = `
Array [
  Object {
    "code": "invalid_type",
    "expected": "string",
    "message": "Expected string, received null",
    "path": Array [
      "detail",
      "metadata",
      "x-emc-ubid",
    ],
    "received": "null",
  },
]
`;

exports[`updateQuantityCT handler when update type is increment should not make requests if quantity is missing 1`] = `
Array [
  Object {
    "code": "invalid_type",
    "expected": "number",
    "message": "Expected number, received null",
    "path": Array [
      "detail",
      "payload",
      "quantity",
    ],
    "received": "null",
  },
]
`;

exports[`updateQuantityCT handler when update type is increment should not make requests if sku is missing 1`] = `
Array [
  Object {
    "code": "invalid_type",
    "expected": "string",
    "message": "Expected string, received null",
    "path": Array [
      "detail",
      "payload",
      "sku",
    ],
    "received": "null",
  },
]
`;

exports[`updateQuantityCT handler when update type is increment should not make requests if updateType is missing 1`] = `
Array [
  Object {
    "code": "invalid_type",
    "expected": "'increment' | 'decrement'",
    "message": "Expected 'increment' | 'decrement', received null",
    "path": Array [
      "detail",
      "payload",
      "updateType",
    ],
    "received": "null",
  },
]
`;
